[17.0][FIX] account: Correct payment amount in wizard when grouping discounted and non-discounted bills

cbeddies opened the pull request:
Description of the issue/feature this PR addresses:
If the user has two invoices/bills selected where one falls within the parameters of a discounted payment and the other does not, the calculated payment total and payment difference will be wrong.

Current behavior before PR:
If I have two Bills, both for $25.00 where one falls within the parameters of a discount for 2% in 10 Days and the other doesn't, and I select both of them and click register payment, and the select "Group Payments" (group_payment), the "Amount" (amount) will be $0.50 and the Payment Difference (payment_difference) will be $49.50. 

Desired behavior after PR is merged:
In the same scenario, I would expect the "Amount" to be $49.50 and the "Payment Difference" to be $0.50


To Reproduce:

- Create Payment Term "2% 10, Net 30" this payment term should be configured to have a discount of 2% if paid in the first 10 days, and then 100% of the bill is due in 30 Days. 
- Create two Purchase Orders (doesn't matter which products you purchase), receive products, and create bills. Make sure one has the new payment term you created and its bill date puts it within the discount parameters. The other bill should either have a different payment term that doesn't have a discount or the same term but not be within the parameters to have the discount apply (bill is older than 10 days in this case)
- Go to Accounting --> Vendors --> Bills and select both of these new bills and click "Register Payment"
- In the wizard, click "Group Payments" and you will see the amount and payment difference are incorrect.

**This process can also be done with sale orders instead, you will get the same result with two invoices for customers.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/165539).

gawa-odoo reviewed (changes requested):
Indeed, there is an issue.
But your fix is not completely right, because it does not take into account credit notes (which is why there is a test failing because of your fix).
You might have to do it by hand, to base it on the fact that we have move of purchase_types

You will also need to add a commit where you sign the CLA https://github.com/odoo/odoo/blob/17.0/doc/cla/sign-cla.md

If not, I will takeover and make the fix myself.

Have a nice day

cbeddies commented:
Hello @gawa-odoo I believe I solved what was failing in test. I have also signed the CLA and squashed my commits and corrected them so that I no longer fail the style check. I am pretty sure all checks should pass but I will check again when the jobs are finished running.

Thanks!

gawa-odoo reviewed (commented):
While advising a change on the computation, I also propose some smallish changes in the test
thanks!

cbeddies commented on addons/account/wizard/account_payment_register.py:
```diff
@@ -505,7 +505,7 @@ def _get_total_amount_using_same_currency(self, batch_result, early_payment_disc
                 mode = 'early_payment'
```
Just tested this, looks like it works well and is safer in the long run so I changed my code to use this.

cbeddies commented on addons/account/tests/test_account_payment_register.py:
```diff
@@ -105,7 +122,25 @@ def setUpClass(cls, chart_template_ref=None):
+            'partner_id': cls.partner_b.id,
+            'invoice_payment_term_id': cls.invoice_payment_term_1.id,
+            'currency_id': cls.currency_data['currency'].id,
+            'invoice_line_ids': [(0, 0, {'product_id': cls.product_a.id, 'price_unit': 25.0, 'tax_ids': []})],
```
This and other lines like it should be adjusted now.

cbeddies commented:
@gawa-odoo I fixed most of the items you pointed out. For some reason my new unit test keeps failing and I am not sure why. It keeps returning with a value of 0 when two 25 dollar bills and a 10 dollar refund are posted and a wizard is created from them. The previous test does exactly what I expect but not this new one. My understanding of unit tests is that the objects created in setup are not changed between tests (always start fresh) is this wrong?

gawa-odoo commented on addons/account/tests/test_account_payment_register.py:
```diff
@@ -115,7 +150,14 @@ def setUpClass(cls, chart_template_ref=None):
+            'move_type': 'in_refund',
+            'date': fields.Date.today(),
+            'invoice_date': fields.Date.today(),
+            'partner_id': cls.partner_a.id,
```
You lack the same currency as the in_invoice associated to be in the same batch. 
Either you add it here, or you remove it from the bills you created. Probably best to remove it from the bills above, as we don't need foreign currency

cbeddies replied:
Yep that would make sense, I will remove from the bills and push back up. Should pass after that.

gawa-odoo reviewed (commented):
Found it :D

gawa-odoo commented on addons/account/tests/test_account_payment_register.py:
```diff
@@ -115,7 +148,14 @@ def setUpClass(cls, chart_template_ref=None):
+            'move_type': 'in_refund',
+            'date': fields.Date.today(),
+            'invoice_date': fields.Date.today(),
+            'partner_id': cls.partner_a.id,
```
Rohh sorry, you also need the same partner.

cbeddies replied:
You are right I don't know how I missed that. Thought they were all partner A, will fix now. Thanks!

gawa-odoo commented:
Perfect! 
Can you squash your commit and put a commit message that will have the part about "Tto reproduce" from you PR's message, and also that the fix was signing the amount?
If not, I would have to use the PR's message (that references often the term PR :smile: )

Thanks

cbeddies commented:
Sorry about that, last couple days were hectic. Should be updated and ready to go now. After the checks run that is

cbeddies commented:
Should be all good to go

gawa-odoo commented:
Thanks!
@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward.
